lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -1078,20 +1080,44 @@ ConstExprFunctionState::computeSingleStoreAddressValue(SILValue addr) {
   SingleValueInstruction *addrInst = dyn_cast<AllocStackInst>(addr);
   if (!addrInst)
     addrInst = dyn_cast<PointerToAddressInst>(addr);
-  if (!addrInst)
+  if (!addrInst) {
+    // Example SIL snippet involving index_addr that we can handle:
+    // %78 = index_addr %73 : $*Int32, %77 : $Builtin.Word // user: %86
+    // function_ref SignedInteger<>.init<A>(_:)
+    // %85 = function_ref @... // user: %86
+    // Here the func call writes to %78.
+    // %86 = apply %85<Int32, Int>(%78, %83, %80) : $@convention(method)
     addrInst = dyn_cast<IndexAddrInst>(addr);
+  }
+  if (!addrInst) {
+    if (auto *teai = dyn_cast<TupleElementAddrInst>(addr)) {
+      // In this case, we simply const-evaluate this inst, which involves first
+      // const-evaluating the tuple operand of this inst (to an aggregate const
+      // value). e.g. %131 below.
+      // %132 = tuple_element_addr %131 : $*(Int32, Int32, Int32, Int32), 0
+      auto tupleEltAddr = computeConstantValue(addr);
+      return tupleEltAddr;
+    }
+  }
   if (!addrInst)
     return evaluator.getUnknown(addr, UnknownReason::Default);
 
+  return computeSingleStoreAddressValueHelper(addrInst);
+}
+
+/// This does the heavy-lifting of computeSingleStoreAddressValue() above, and
+/// may recursively call itself when const-evaluating TupleElementAddrInst.
+SymbolicValue
+ConstExprFunctionState::computeSingleStoreAddressValueHelper(SILValue addr) {
```

Comment:
May not pertinent to this CL, but would this be better implemented as a SILValue/SILInstructionVisitor?

Reply:
Yes this also came up in some past PRs from @lattner on const eval.

Reply:
This function seems highly redundant with the existing getConstant() support for addresses.  Can we reuse it?